### PR TITLE
community access: manage access 

### PR DESCRIPTION
**Pull request by mitsosf**
closes https://github.com/zenodo/zenodo-rdm/issues/250

### Description

Manages community access status.

If the user is an admin, the new config variable will have no effect on the user's UX.
If it is a normal user, they are not allowed to change the access rights on the backend and also not able to see the options on the frontend.

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [x] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [x] I've added relevant test cases.
- [x] I've added relevant documentation.
- [x] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code).
- [x] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code).
- [x] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code).
- [x] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [x] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [x] I've NOT included third-party code (copy/pasted source code or new dependencies).

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_communities/communities/services/components.py [hunk 1]
```diff
@@ -70,10 +70,18 @@ class CommunityAccessComponent(ServiceComponent):
...
-            # populate the record's access field with the data already
-            # validated by marshmallow
+            access = data.get("access", {})
+            new_record_access = access.get("visibility")
+            if record.access.visibility != new_record_access:
+                can_manage_access = self.service.check_permission(
+                    identity, "manage_access", record=record
+                )
+                if not can_manage_access:
+                    if "visibility" in access:
```
Comment (kpsherva) [inline invenio_communities/communities/services/components.py:80]:
```suggestion
                if not can_manage_access and "visibility" in access:
```

File: invenio_communities/communities/services/components.py [hunk 2]
```diff
@@ -70,10 +70,18 @@ class CommunityAccessComponent(ServiceComponent):
     def _populate_access_and_validate(self, identity, data, record, **kwargs):
         """Populate and validate the community's access field."""
         if record is not None and "access" in data:
-            # populate the record's access field with the data already
-            # validated by marshmallow
+            access = data.get("access", {})
+            new_record_access = access.get("visibility")
```
Comment (kpsherva) [inline invenio_communities/communities/services/components.py:74]:
```suggestion
            access = data.get("access", {}).get("visibility")
```

File: invenio_communities/generators.py [hunk 3]
```diff
@@ -158,6 +159,9 @@ def communities(self, identity):
 
     def needs(self, record=None, community_id=None, **kwargs):
         """Enabling Needs."""
+        if record is None:
```
Comment (kpsherva) [inline invenio_communities/generators.py:162]: why was this needed?

File: invenio_communities/views/communities.py [hunk 4]
```diff
@@ -224,9 +236,16 @@ def communities_requests(pid_value, community, community_ui):
...
+            "update",
+            "read",
+            "search_requests",
+            "search_invites",
+            "manage_access",
+        ]
     )
-    if not permissions["can_update"]:
+
+    if not permissions["can_manage_access"]:
```
Comment (kpsherva) [inline invenio_communities/views/communities.py:248]: this might be breaking the tests

**Comment by kpsherva** [comment]
api tests for the PR are missing (unless they are not pushed yet)

File: invenio_communities/communities/services/components.py (hunk 2 above)
Comment (mitsosf) [reply invenio_communities/communities/services/components.py:79]: I am using the access variable further down, but can modify this part if we decide to raise an exception, as suggested later.

File: invenio_communities/communities/services/components.py [hunk 5]
```diff
@@ -70,10 +70,18 @@ class CommunityAccessComponent(ServiceComponent):
...
+            if record.access.visibility != new_record_access:
+                can_manage_access = self.service.check_permission(
```
[the remark this answers is not in the document]
Comment (mitsosf) [reply invenio_communities/communities/services/components.py]: I'm basically copying the logic from Alex's PR for records, I thought he must have had his reasons to go this way.

File: invenio_communities/generators.py (hunk 3 above)
Comment (mitsosf) [reply invenio_communities/generators.py]: I removed it now, I placed it there to try and use this function also when creating a new community. We have this implementation elsewhere in the code, checking if the object being passed is empty, hence new.
Now it is not required, because I am using a different permission for creating a community and mofdifying one.

**Comment by kpsherva** [comment]
tested with REST api call, 
```
 curl -X POST -k -H "Authorization: Bearer <token>" -H "Content-Type: application/json" https://127.0.0.1:5000/api/communities 
 --data '{"access": {"visibility": "restricted"}, "slug": "abs-restricted", "metadata": {"title": "abc-test"}}'
```
the community is created with public access, I would expect it raises an error instead of creating public community. Ping @slint to confirm if thats what you agreed on

File: invenio_communities/communities/services/components.py [hunk 6]
```diff
@@ -69,12 +69,25 @@ class CommunityAccessComponent(ServiceComponent):
...
+        access = data.get("access", {})
+        new_visibility = access.get("visibility")
+
+        record_has_defined_access = "access" in record
+        is_restricting = new_visibility == "restricted"
+
+        action = "manage_access" if record_has_defined_access else "create_restricted"
+
+        if record.access.visibility != new_visibility and is_restricting:
+            self.service.require_permission(identity, action, record=record)
```
Comment (slint) [inline invenio_communities/communities/services/components.py:86]:
```suggestion
        record_has_defined_access = "access" in record

        action = "manage_access" if record_has_defined_access else "create_restricted"

        if record.access.visibility != new_visibility:
            self.service.require_permission(identity, action, record=record)
```

minor: I wouldn't limit only to restricting, but in general to all changes. 

Imagine e.g. an admin making a community private (because it's potentially spam/malicious), and then the community owner switching it back to public visibility.

File: invenio_communities/views/communities.py [hunk 7]
```diff
@@ -150,6 +150,11 @@ def communities_new():
     can_create = current_communities.service.check_permission(g.identity, "create")
     if not can_create:
         raise PermissionDeniedError()
+
+    can_create_restricted = current_communities.service.check_permission(
```
Comment (slint) [inline invenio_communities/views/communities.py:154]: minor: shouldn't this be just `can_manage_access` again? I know it's not very explicit, but do we want to actually separate the two actions? I.e. is there a scenario where we allow a user to create restricted communities, but not change the visibility afterwards?

**Review by slint — COMMENTED** [review]
Some comments on reducing the logic, but I also might be missing some previous discussions, so feel free to dismiss them if not applicable 😅
